### PR TITLE
Add policy for disconnecting request body halfway.

### LINKED ISSUE #1755: MockWebServer SocketPolicy interrupting request
SocketPolicy has DISCONNECT_AFTER_REQUEST. Is there a way to force the socket to disconnect during the request, i.e. while the request is being read but before it is finished?

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #1755.

### CODE REVIEW

**Comment by swankjesse** [comment]
LGTM!

**Comment by swankjesse** [comment]
```
Tests run: 22, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 4.305 sec <<< FAILURE! - in com.squareup.okhttp.mockwebserver.MockWebServerTest

disconnectRequestHalfway(com.squareup.okhttp.mockwebserver.MockWebServerTest)  Time elapsed: 0.094 sec  <<< FAILURE!

java.lang.AssertionError: null

	at org.junit.Assert.fail(Assert.java:86)

	at org.junit.Assert.assertTrue(Assert.java:41)

	at org.junit.Assert.assertTrue(Assert.java:52)

	at com.squareup.okhttp.mockwebserver.MockWebServerTest.disconnectRequestHalfway(MockWebServerTest.java:279)
```

**Comment by JakeWharton** [comment]
Ouch. So much for my in-code comment.

Any ideas on how to make failure more deterministic here?

**Comment by swankjesse** [comment]
Go for a full 1 MiB request body. You need to fill some big buffers.

**Comment by JakeWharton** [comment]
I did 1MB chunks in a 1GB body. It seems extremely deterministic on my machine, but we'll see what Travis things.

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/SocketPolicy.java [hunk 1]
```diff
@@ -50,6 +50,9 @@
    */
   DISCONNECT_AFTER_REQUEST,
 
+  /** Close connection after reading half of the request body (if present). */
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/SocketPolicy.java]: Should I add a warning about the asynchronous notification of close which might not result in failure on small bodies?

Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/SocketPolicy.java]: Wouldn't hurt! This entire class is so painfully organic in design. I hope to someday make a do-over where we borrow models from the client.

**Comment by hey99xx** [comment]
The 10f value failed on my pipeline. Is it expected to fail nondeterministically? 15f seems to run a bit more consistently but still failed occasionally.

**Pull request merged**